app/reducers/reducer.ts

```diff
@@ -827,6 +845,23 @@ function hasSelection<T extends GridInterfaceState>(state: T): boolean {
   return isBuilderState(state) && hasMultipleCells(state.selection);
 }
 
+function pushToHistory<T extends GridInterfaceState>(state: T): T {
+  if (!isBuilderState(state)) {
+    return state;
+  }
+
+  const MAX_HISTORY_LENGTH = 20;
```

Comment:
Total edge case (I'm OK w/ a "won't fix") but do you think we should de-dupe the grid state before pushing? To reproduce you can type over existing cells w/ a matching cell and then the first undos will actually be no-ops. Again I think this is OK as it's a weird action to take as a user, particularly if de-duping is non-trivial.

Reply:
This was pretty easy to add, and should be easy to remove if it proves unwanted! FWIW I tested in a few places and it seem like most applications don't dedupe.